public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts

```diff
@@ -121,18 +109,43 @@ export function createSceneObjectsForPanels(oldPanels: PanelModel[]): SceneGridI
 
   // commit a row if it's the last one
   if (currentRow) {
-    panels.push(
-      new SceneGridRow({
-        title: currentRow!.title,
-        y: currentRow.gridPos.y,
-        children: currentRowPanels,
-      })
-    );
+    panels.push(createRowFromPanelModel(currentRow, currentRowPanels));
   }
 
   return panels;
 }
 
+function createRowFromPanelModel(row: PanelModel, content: SceneGridItemLike[]): SceneGridItemLike {
+  if (Boolean(row.collapsed)) {
+    if (row.panels) {
+      content = row.panels.map(buildGridItemForPanel);
+    }
+  }
+
+  let behaviors: SceneObject[] | undefined;
+  let children = content;
+
+  if (row.repeat) {
+    // For repeated rows the children are stored in the behavior
+    children = [];
+    behaviors = [
+      new RowRepeaterBehavior({
+        variableName: row.repeat,
+        sources: content,
```

Comment:
Why do we need to pass the sources to the behavior? Think it could figure out what to repeat internally, given it should be attached to `SceneGridRow` only?

Reply:
Ah I think I understand why. Basically, because you are altering the children of a row as a result of processing repeats?

Reply:
@dprokop 

yea, I keeping them separate here so I do not render them until they have been "repeated" and wrapped in the correct SceneVariableSet yet (for local variable scope). it does simplify the logic a bit. But will make editing a bit tricker, I have to take state changes done on the row children and copy those state changes into the corresponding source item. So maybe will revisit this design decision if that turns out to be messy.

Reply:
@dprokop 

yea, I keeping them separate here so I do not render them until they have been "repeated" and wrapped in the correct SceneVariableSet yet (for local variable scope). it does simplify the logic a bit. But will make editing a bit tricker, I have to take state changes done on the row children and copy those state changes into the corresponding source item. So maybe will revisit this design decision if that turns out to be messy.